src/Microsoft.DotNet.Darc/DarcLib/GitHubClient.cs

```diff
@@ -1345,4 +1300,63 @@ public async Task<List<string>> GetPullRequestCommentsAsync(string pullRequestUr
         
         return comments.Select(comment => comment.Body).ToList();
     }
+
+    /// <summary>
+    /// This method fills a functionality that's currently missing from Octokit: fetching Github resources using eTags.
+    /// eTags allow us to cache mutable resources and efficiently check if the resource has changed on Github since the last fetch.
+    /// </summary>
+    /// <typeparam name="T">The domain class of the resource in our server</typeparam>
+    /// <typeparam name="K">The class of the resource in Octokit</typeparam>
+    /// <param name="resourceKey">The key used to cache the resource in redis</param>
+    /// <param name="resourceUri">The uri used to request the resource from Github</param>
+    /// <param name="client">The github client that makes the request</param>
+    /// <param name="resourceConverter">Function to convert the resource from Octokit to our domain class</param>
+    /// <returns>The resource of type T</returns>
+    /// <exception cref="DarcException"></exception>
+    protected virtual async Task<T> RequestResourceUsingEtagsAsync<T, K>(
+        string resourceKey,
+        Uri resourceUri,
+        IGitHubClient client,
+        Func<K, T> resourceConverter)
+        where T : class, IGithubEtagResource
+    {
+        var cachedResource = await _cache.TryGetAsync<T>(resourceKey);
+        string? entityTag = cachedResource?.Etag;
+        var headers = new Dictionary<string, string>
+        {
+            { "Accept", "application/vnd.github.v3+json" },
+        };
+        if (entityTag != null)
+        {
+            headers.Add("If-None-Match", entityTag);
+        }
+        var response = await client.Connection.Get<K>(resourceUri, headers);
+        if (response.HttpResponse.StatusCode == HttpStatusCode.NotModified && cachedResource != null)
+        {
+            // TODO: Add telemetry for cache hits to measure the impact of this optimization.
+            return cachedResource;
+        }
+        else
+        {
+            if (response.HttpResponse.StatusCode == HttpStatusCode.OK)
+            {
+                string? etag = response.HttpResponse.Headers
+                    .FirstOrDefault(h => h.Key.Equals("Etag", StringComparison.OrdinalIgnoreCase))
+                    .Value;
+
+                var resource = resourceConverter(response.Body);
+
+                if (etag != null)
+                {
+                    resource.Etag = etag;
+                    await _cache.TrySetAsync<T>(resourceKey, resource);
```

Comment:
How about expiration? I think we should keep it low since we are not actively deleting the data anywhere